core/src/main/java/hudson/slaves/SlaveComputer.java

```diff
@@ -482,7 +481,15 @@ public void onClosed(Channel c, IOException cause) {
                     taskListener.getLogger().println("Connection terminated");
                 }
                 closeChannel();
-                launcher.afterDisconnect(SlaveComputer.this, taskListener);
+                try {
+                    launcher.afterDisconnect(SlaveComputer.this, taskListener);
+                } catch (Throwable t) {
+                    LogRecord lr = new LogRecord(Level.SEVERE,
+                            "Launcher {0}'s afterDisconnect method propagated an exception when {1}'s connection was closed: {2}");
```

Comment:
Printing the `ComputerLauncher.toString` is probably senseless since any interesting information about the implementation would be in the stack trace, and `toString` is unlikely to be overridden from `Object`. Printing the `SlaveComputer.toString` is definitely senseless since this is not overridden; you want to use `nodeName` (or indirectly via `getName`).